src/btree/bt_cursor.c

```diff
@@ -1360,9 +1365,19 @@ __cursor_chain_exceeded(WT_CURSOR_BTREE *cbt)
 	    page->modify != NULL && page->modify->mod_row_update != NULL)
 		upd = page->modify->mod_row_update[cbt->slot];
 
-	for (i = 0; upd != NULL; ++i, upd = upd->next) {
+	for (i = 0, upd_size = 0; upd != NULL; ++i, upd = upd->next) {
 		if (WT_UPDATE_DATA_VALUE(upd))
 			return (false);
+		upd_size += WT_UPDATE_MEMSIZE(upd);
+		if (upd_size >= WT_MODIFY_MEM_FACTOR * cursor->value.size)
+			return (true);
+		/*
+		 * When history has to be maintained, creating extra copies
+		 * of large documents multiplies cache pressure, so avoid it,
+		 * by growing the modify chain.
+		 */
+		if (!__wt_txn_upd_visible_all(session, upd))
```

Comment:
I think this check better either 'before previous if check (line 1372)' or 'after next if check (line 1381)'. In between seems to be a bit odd, as both if checks are in way to constrain the length of the update chain. @michaelcahill, request for your opinion.

Reply:
@bvpvamsikrishna, I've put it this way deliberately, maybe a comment can help?

Basically, the size check means we don't create huge amounts of cache pressure (10x 8 byte updates to a 10MB record creates 125000x cache pressure today).  The size check keeps the cache pressure to 2x.

The length check is mostly about read performance: we don't want to do too many apply operations on every read.

The reason to put the visibility check in between is to keep caching small updates when they can't be cleaned up.  It's okay to create new complete records when we can free up the last one.

Does that make sense, could I add a comment to clarify?

Reply:
@michaelcahill, Thanks for clarification. sounds good. 
Only thing is comment is not moved with the if check. I can move it, if we plan to merge this soon.